File: tests/PenaltyFunctionTests.cpp
#include <gtest/gtest.h>

#include "TestPenaltyFunctions.hpp"

using namespace clf;

TEST(PenaltyFunctionTests, DenseTest0) {
  // create the example penalty function
  tests::DensePenaltyFunctionTest0 func;

  // the input and output dimensions are hardcoded as 3 and 2, respectively
  EXPECT_EQ(func.InputDimension(), 3);
  EXPECT_EQ(func.OutputDimension(), 2);

  // evaluate the cost function 
  const Eigen::VectorXd beta = Eigen::VectorXd::Random(3);
  const Eigen::VectorXd eval = func.Evaluate(beta);
  EXPECT_EQ(eval.size(), func.OutputDimension());
  EXPECT_NEAR(eval(0), beta(0), 1.0e-14);
  EXPECT_NEAR(eval(1), beta(0)*(1.0-beta(2)), 1.0e-14);

  const Eigen::MatrixXd exactJac = func.Jacobian(beta);
  EXPECT_EQ(exactJac.rows(), 2);
  EXPECT_EQ(exactJac.cols(), 3);  
  const Eigen::MatrixXd fdJac = func.JacobianFD(beta);
  EXPECT_EQ(exactJac.rows(), fdJac.rows());
  EXPECT_EQ(exactJac.cols(), fdJac.cols());
  EXPECT_NEAR((exactJac-fdJac).norm(), 0.0, 1.0e-13);

  const Eigen::MatrixXd weights = Eigen::VectorXd::Random(2);
  const Eigen::MatrixXd exactHess = func.Hessian(beta, weights);
  EXPECT_EQ(exactHess.rows(), 3);
  EXPECT_EQ(exactHess.cols(), 3);
  const Eigen::MatrixXd fdHess = func.HessianFD(beta, weights);
  EXPECT_EQ(exactHess.rows(), fdHess.rows());
  EXPECT_EQ(exactHess.cols(), fdHess.cols());
  EXPECT_NEAR((exactHess-fdHess).norm(), 0.0, 1.0e-13);
}

TEST(PenaltyFunctionTests, DenseTest1) {
  // create the example penalty function
  tests::DensePenaltyFunctionTest1 func;

  // the input and output dimensions are hardcoded as 3 and 2, respectively
  EXPECT_EQ(func.InputDimension(), 3);
  EXPECT_EQ(func.OutputDimension(), 6);

  // evaluate the cost function 
  const Eigen::VectorXd beta = Eigen::VectorXd::Random(3);
  const Eigen::VectorXd eval = func.Evaluate(beta);
  EXPECT_EQ(eval.size(), func.OutputDimension());
  EXPECT_NEAR(eval(0), 1.0-beta(1), 1.0e-14);
  EXPECT_NEAR(eval(1), 1.0-beta(1)+beta(2), 1.0e-14);
  EXPECT_NEAR(eval(2), beta(2), 1.0e-14);
  EXPECT_NEAR(eval(3), beta(2)*(1.0-beta(1)), 1.0e-14);
  EXPECT_NEAR(eval(4), beta(0)*beta(2), 1.0e-14);
  EXPECT_NEAR(eval(5), beta(0)*beta(0)*beta(1), 1.0e-13);

  const Eigen::MatrixXd exactJac = func.Jacobian(beta);
  EXPECT_EQ(exactJac.rows(), 6);
  EXPECT_EQ(exactJac.cols(), 3);  
  const Eigen::MatrixXd fdJac = func.JacobianFD(beta);
  EXPECT_EQ(exactJac.rows(), fdJac.rows());
  EXPECT_EQ(exactJac.cols(), fdJac.cols());
  EXPECT_NEAR((exactJac-fdJac).norm(), 0.0, 1.0e-13);

  const Eigen::MatrixXd weights = Eigen::VectorXd::Random(6);
  const Eigen::MatrixXd exactHess = func.Hessian(beta, weights);
  EXPECT_EQ(exactHess.rows(), 3);
  EXPECT_EQ(exactHess.cols(), 3);
  const Eigen::MatrixXd fdHess = func.HessianFD(beta, weights);
  EXPECT_EQ(exactHess.rows(), fdHess.rows());
  EXPECT_EQ(exactHess.cols(), fdHess.cols());
  EXPECT_NEAR((exactHess-fdHess).norm(), 0.0, 1.0e-13);
}

TEST(PenaltyFunctionTests, SparseTest0) {
  // create the example penalty function
  tests::SparsePenaltyFunctionTest0 func;

  // the input and output dimensions are hardcoded as 3 and 2, respectively
  EXPECT_EQ(func.InputDimension(), 3);
  EXPECT_EQ(func.OutputDimension(), 2);

  // evaluate the cost function 
  const Eigen::VectorXd beta = Eigen::VectorXd::Random(3);
  const Eigen::VectorXd eval = func.Evaluate(beta);
  EXPECT_EQ(eval.size(), func.OutputDimension());
  EXPECT_NEAR(eval(0), beta(0), 1.0e-14);
  EXPECT_NEAR(eval(1), beta(0)*(1.0-beta(2)), 1.0e-14);

  const Eigen::SparseMatrix<double> exactJac = func.Jacobian(beta);
  EXPECT_EQ(exactJac.rows(), 2);
  EXPECT_EQ(exactJac.cols(), 3);  
  const Eigen::SparseMatrix<double> fdJac = func.JacobianFD(beta);
  EXPECT_EQ(exactJac.rows(), fdJac.rows());
  EXPECT_EQ(exactJac.cols(), fdJac.cols());
  EXPECT_NEAR((exactJac-fdJac).norm(), 0.0, 1.0e-13);

  const Eigen::VectorXd weights = Eigen::VectorXd::Random(2);
  const Eigen::SparseMatrix<double> exactHess = func.Hessian(beta, weights);
  EXPECT_EQ(exactHess.rows(), 3);
  EXPECT_EQ(exactHess.cols(), 3);
  const Eigen::SparseMatrix<double> fdHess = func.HessianFD(beta, weights);
  EXPECT_EQ(exactHess.rows(), fdHess.rows());
  EXPECT_EQ(exactHess.cols(), fdHess.cols());
  EXPECT_NEAR((exactHess-fdHess).norm(), 0.0, 1.0e-13);
}

TEST(PenaltyFunctionTests, SparseTest1) {
  // create the example penalty function
  tests::SparsePenaltyFunctionTest1 func;

  // the input and output dimensions are hardcoded as 3 and 2, respectively
  EXPECT_EQ(func.InputDimension(), 3);
  EXPECT_EQ(func.OutputDimension(), 6);

  // evaluate the cost function 
  const Eigen::VectorXd beta = Eigen::VectorXd::Random(3);
  const Eigen::VectorXd eval = func.Evaluate(beta);
  EXPECT_EQ(eval.size(), func.OutputDimension());
  EXPECT_NEAR(eval(0), 1.0-beta(1), 1.0e-14);
  EXPECT_NEAR(eval(1), 1.0-beta(1)+beta(2), 1.0e-14);
  EXPECT_NEAR(eval(2), beta(2), 1.0e-14);
  EXPECT_NEAR(eval(3), beta(2)*(1.0-beta(1)), 1.0e-14);
  EXPECT_NEAR(eval(4), beta(0)*beta(2), 1.0e-14);
  EXPECT_NEAR(eval(5), beta(0)*beta(0)*beta(1), 1.0e-14);

  const Eigen::SparseMatrix<double> exactJac = func.Jacobian(beta);
  EXPECT_EQ(exactJac.rows(), 6);
  EXPECT_EQ(exactJac.cols(), 3);  
  const Eigen::SparseMatrix<double> fdJac = func.JacobianFD(beta);
  EXPECT_EQ(exactJac.rows(), fdJac.rows());
  EXPECT_EQ(exactJac.cols(), fdJac.cols());
  EXPECT_NEAR((exactJac-fdJac).norm(), 0.0, 1.0e-13);

  const Eigen::VectorXd weights = Eigen::VectorXd::Random(6);
  const Eigen::SparseMatrix<double> exactHess = func.Hessian(beta, weights);
  EXPECT_EQ(exactHess.rows(), 3);
  EXPECT_EQ(exactHess.cols(), 3);
  const Eigen::SparseMatrix<double> fdHess = func.HessianFD(beta, weights);
  EXPECT_EQ(exactHess.rows(), fdHess.rows());
  EXPECT_EQ(exactHess.cols(), fdHess.cols());
  EXPECT_NEAR((exactHess-fdHess).norm(), 0.0, 1.0e-12);
}

File: tests/PenaltyFunctionTests.py
"""! @brief Test the penalty function (clf::PenaltyFunction) python interface"""

import unittest

import numpy as np
import scipy

import random

from TestPenaltyFunctions import *
import PyCoupledLocalFunctions as clf

class TestDensePenaltyFunction(unittest.TestCase):
    """! Test the dense penalty function (clf::DensePenaltyFunction) python interface"""
    def tearDown(self):
        """! Compare the derivative information to finite difference approximations"""
        
        self.assertEqual(self.func.indim, 3)
        self.assertEqual(self.func.outdim, self.outdim)

        beta = np.array([random.uniform(-1.0, 1.0) for i in range(3)])

        jac = self.func.Jacobian(beta)
        self.assertEqual(np.shape(jac) [0], self.outdim)
        self.assertEqual(np.shape(jac) [1], 3)
        jacFD = self.func.JacobianFD(beta)
        self.assertEqual(np.shape(jacFD) [0], self.outdim)
        self.assertEqual(np.shape(jacFD) [1], 3)
        self.assertAlmostEqual(np.linalg.norm(jac-jacFD), 0.0)
        
        weights = np.array([random.uniform(-1.0, 1.0) for i in range(self.outdim)])
        hess = self.func.Hessian(beta, weights)
        self.assertEqual(np.shape(hess) [0], 3)
        self.assertEqual(np.shape(hess) [1], 3)
        hessFD = self.func.HessianFD(beta, weights)
        self.assertEqual(np.shape(hessFD) [0], 3)
        self.assertEqual(np.shape(hessFD) [1], 3)
        self.assertAlmostEqual(np.linalg.norm(hess-hessFD), 0.0)
    
    def test_dense_test0(self):
        """! Check TestPenaltyFunctions.DensePenaltyFunctionTest0"""

        ## The output dimension
        self.outdim = 2
        ## The penalty function
        self.func = DensePenaltyFunctionTest0()
        
        # evaluate the cost function 
        beta = np.array([random.uniform(-1.0, 1.0) for i in range(3)])
        fx = self.func.Evaluate(beta)
        self.assertEqual(len(fx), self.func.outdim);
        self.assertAlmostEqual(fx[0], beta[0])
        self.assertAlmostEqual(fx[1], beta[0]*(1.0-beta[2]))

    def test_dense_test1(self):
        """! Check TestPenaltyFunctions.DensePenaltyFunctionTest1"""
        
        ## The output dimension
        self.outdim = 6
        ## The penalty function
        self.func = DensePenaltyFunctionTest1()
        
        # evaluate the cost function 
        beta = np.array([random.uniform(-1.0, 1.0) for i in range(3)])
        fx = self.func.Evaluate(beta)
        self.assertEqual(len(fx), self.func.outdim);
        self.assertAlmostEqual(fx[0], 1.0-beta[1])
        self.assertAlmostEqual(fx[1], 1.0-beta[1]+beta[2])
        self.assertAlmostEqual(fx[2], beta[2])
        self.assertAlmostEqual(fx[3], beta[2]*(1.0-beta[1]))
        self.assertAlmostEqual(fx[4], beta[0]*beta[2])
        self.assertAlmostEqual(fx[5], beta[0]*beta[0]*beta[1])

class TestSparsePenaltyFunction(unittest.TestCase):
    """! Test the sparse penalty function (clf::SparsePenaltyFunction) python interface"""
    def tearDown(self):
        """! Compare the derivative information to finite difference approximations"""
                
        self.assertEqual(self.func.indim, 3)
        self.assertEqual(self.func.outdim, self.outdim)

        beta = np.array([random.uniform(-1.0, 1.0) for i in range(3)])

        jac = self.func.Jacobian(beta)
        self.assertEqual(np.shape(jac) [0], self.outdim)
        self.assertEqual(np.shape(jac) [1], 3)
        jacFD = self.func.JacobianFD(beta)
        self.assertEqual(np.shape(jacFD) [0], self.outdim)
        self.assertEqual(np.shape(jacFD) [1], 3)
        self.assertAlmostEqual(scipy.sparse.linalg.norm(jac-jacFD), 0.0)
        
        weights = np.array([random.uniform(-1.0, 1.0) for i in range(self.outdim)])
        hess = self.func.Hessian(beta, weights)
        self.assertEqual(np.shape(hess) [0], 3)
        self.assertEqual(np.shape(hess) [1], 3)
        hessFD = self.func.HessianFD(beta, weights)
        self.assertEqual(np.shape(hessFD) [0], 3)
        self.assertEqual(np.shape(hessFD) [1], 3)
        self.assertAlmostEqual(scipy.sparse.linalg.norm(hess-hessFD), 0.0)

    def test_sparse_test0(self):
        """! Check TestPenaltyFunctions.SparsePenaltyFunctionTest0"""

        ## The output dimension
        self.outdim = 2
        ## The penalty function
        self.func = SparsePenaltyFunctionTest0()
        
        # evaluate the cost function 
        beta = np.array([random.uniform(-1.0, 1.0) for i in range(3)])
        fx = self.func.Evaluate(beta)
        self.assertEqual(len(fx), self.func.outdim);
        self.assertAlmostEqual(fx[0], beta[0])
        self.assertAlmostEqual(fx[1], beta[0]*(1.0-beta[2]))

    def test_sparse_test1(self):
        """! Check TestPenaltyFunctions.SparsePenaltyFunctionTest0"""

        ## The output dimension
        self.outdim = 6
        ## The penalty function
        self.func = SparsePenaltyFunctionTest1()
        
        # evaluate the cost function 
        beta = np.array([random.uniform(-1.0, 1.0) for i in range(3)])
        fx = self.func.Evaluate(beta)
        self.assertEqual(len(fx), self.func.outdim);
        self.assertAlmostEqual(fx[0], 1.0-beta[1])
        self.assertAlmostEqual(fx[1], 1.0-beta[1]+beta[2])
        self.assertAlmostEqual(fx[2], beta[2])
        self.assertAlmostEqual(fx[3], beta[2]*(1.0-beta[1]))
        self.assertAlmostEqual(fx[4], beta[0]*beta[2])
        self.assertAlmostEqual(fx[5], beta[0]*beta[0]*beta[1])

    
